src/proofs/ProofSteps.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2023, Matias Barandiaran <matias.barandiaran03@gmail.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#ifndef GOLEM_ALETHETERMS_H
+#define GOLEM_ALETHETERMS_H
+#include "Term.h"
+#include "Witnesses.h"
+#include "graph/ChcGraph.h"
+#include "utils/SmtSolver.h"
+#include <memory>
+#include <utility>
+
+class Step {
+public:
+    enum stepType { ASSUME, STEP, ANCHOR };
+
+private:
+    int stepId;
+    stepType type;
+    std::vector<std::shared_ptr<Term>> clause;
+    std::string rule;
+    std::vector<int> premises;
+    std::vector<std::pair<std::string, std::string>> args;
+
+public:
+    Step(int stepId, stepType type, std::vector<std::shared_ptr<Term>> clause, std::string rule,
+         std::vector<int> premises)
+        : type(type), stepId(stepId), clause(std::move(clause)), rule(std::move(rule)), premises(std::move(premises)) {}
+    Step(int stepId, stepType type, std::vector<std::shared_ptr<Term>> clause, std::string rule,
+         std::vector<std::pair<std::string, std::string>> args)
+        : type(type), stepId(stepId), clause(std::move(clause)), rule(std::move(rule)), args(std::move(args)) {}
+    Step(int stepId, stepType type, std::vector<std::shared_ptr<Term>> clause, std::string rule)
+        : type(type), stepId(stepId), clause(std::move(clause)), rule(std::move(rule)) {}
+    Step(int stepId, stepType type, std::vector<std::shared_ptr<Term>> clause)
+        : type(type), stepId(stepId), clause(std::move(clause)), rule(" ") {}
+    Step(int stepId, stepType type, std::string rule, std::vector<int> premises)
+        : type(type), stepId(stepId), rule(std::move(rule)), premises(std::move(premises)) {}
+    std::string printStepAlethe();
+    std::string printStepIntermediate();
+};
+
+class Observer {
+public:
+    virtual void update(Step & step) = 0;
+};
+
+class AlethePrintObserver : public Observer {
+    std::ostream & out;
+
+public:
+    explicit AlethePrintObserver(std::ostream & out) : out(out) {}
+    void update(Step & step) override { out << step.printStepAlethe(); }
+};
+
+class IntermediatePrintObserver : public Observer {
+    std::ostream & out;
+
+public:
+    explicit IntermediatePrintObserver(std::ostream & out) : out(out) {}
+    void update(Step & step) override { out << step.printStepIntermediate(); }
+};
+
+class StepHandler {
+
+    InvalidityWitness::Derivation derivation;
+    std::vector<std::shared_ptr<Term>> originalAssertions;
+    Normalizer::Equalities const & normalizingEqualities;
+    std::ostream & out;
+    Logic & logic;
+    ChcDirectedHyperGraph originalGraph;
+
+    std::vector<Observer *> observers;
+
+    int currStep = 0;
+    int transitivityStep;
```

Comment:
Should also be initialized as the other members?

Reply:
It doesn't really need  a default initialization.